README.md
# weather_app

Get today&#x27;s and forcast weather data with the help of this beautiful app built using flutter.\
(*Flutter v1.9.1+hotfix.5 or higher required*)

## APIs used

- [Current weather data](https://openweathermap.org/current)
- [5 day weather forecast](https://openweathermap.org/forecast5)

Free OpenWeatherMap API key has been used in this project.\
If you wish to use your own, edit the "appId" variable in lib/utilities.dart.



lib/change_log.txt
20-Oct-2019 : Moved FutureBuilder code form separate external widget to the body of Home class to allow setState for the refreshing thing

lib/utilities.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:shared_preferences/shared_preferences.dart';

final appId = "499093abb9ae71e744766738f864d7d6"; //API key for OpenWeatherMap
String defaultLocation = "Nagercoil";
//Date format func

String unixToNormal(int unixTime){
  DateTime date = new DateTime.fromMillisecondsSinceEpoch(unixTime*1000).toUtc();
  var fmt=new DateFormat();
  var formattedDate=fmt.format(date);
  return formattedDate;
}

//Useful design functions

textColor() {
  return Colors.white;
}

appBarColor(){
  return Colors.black26;
}

containerTextStyle() {
  return TextStyle(
    fontWeight: FontWeight.w300,
    color: textColor(),
  );
}

containerBoxDecoration() {
  return BoxDecoration(
      color: Colors.black26, borderRadius: BorderRadius.circular(10));
}

moreInfoTextStyle(){
  return TextStyle(
    fontSize: 18,
    fontWeight: FontWeight.w300,
    color: textColor(),
  );
}

//SharedPreferences stuff

setDefaultLocation(String name)async{
  SharedPreferences location= await SharedPreferences.getInstance();
  location.setString("location", name);
}



lib/ui/forecast.dart
// FORECAST
//https://www.countryflags.io/ used for Country icons in Forecast page

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import '../utilities.dart' as utilities;
import 'home.dart';

class Forecast extends StatefulWidget {
  final String location;
  Forecast({Key key, @required this.location}): super(key:key);
  @override
  _ForecastState createState() => _ForecastState(location: location);
}

class _ForecastState extends State<Forecast> {
  final String location;
  var forecastInfo;
  _ForecastState({Key key, @required this.location});
  void initState(){
    super.initState();
    forecastInfo=getForecastInfo(utilities.appId, location);
  }
  refresh(String loco){
    print("From refresh fn-$loco");
    setState(() {
      forecastInfo=null;
      forecastInfo=getForecastInfo(utilities.appId, loco);
    });
  }
  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: forecastInfo,
      builder: (BuildContext context, AsyncSnapshot<Map> snapshot){
    if (snapshot.hasData && snapshot.data.containsKey("list")) {
        return Container(
          decoration: BoxDecoration(
            image: DecorationImage(
              image: NetworkImage(
                  "https://images.pexels.com/photos/2516653/pexels-photo-2516653.jpeg?h=1366&w=768"),
              fit: BoxFit.cover,
            ),
          ),
          child: Scaffold(
            backgroundColor: Colors.black26,
            appBar: AppBar(
              backgroundColor: utilities.appBarColor(),
              title: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  Text(
                    "Forecast",
                    style: utilities.containerTextStyle(),
                  ),
                  FlatButton(
                    child: Icon(Icons.refresh,color: utilities.textColor(),),
                    onPressed:(){
                      refresh(location);
                    },
                  )
                ],
              ),
            ),
            body: ListView(
              //scrollDirection: Axis.horizontal,
              children: <Widget>[
                //Text(snapshot.data.toString(),style: utilities.moreInfoTextStyle(),textAlign: TextAlign.justify,),
                Container(
                  padding: EdgeInsets.all(5),
                  alignment: Alignment.center,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      Image.network("https://www.countryflags.io/${snapshot.data["city"]["country"].toString()}/flat/32.png"),
                      Text(snapshot.data["city"]["name"].toString(),style: utilities.moreInfoTextStyle().merge(TextStyle(fontSize: 35)),),
                      Padding(padding: EdgeInsets.all(5),),
                      Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: <Widget>[
                          Text("Population : ${snapshot.data["city"]["population"].toString()}",style: utilities.moreInfoTextStyle(),),
                          Text("Timezone : ${snapshot.data["city"]["timezone"].toString()}",style: utilities.moreInfoTextStyle(),),
                        ],
                      ),
                      Padding(padding: EdgeInsets.all(20),),
                      Text("5 days/3 hours forecast data",style: utilities.moreInfoTextStyle(),),
                    ],
                  ),
                ),
                Padding(padding: EdgeInsets.all(10),),
                Container(
                  height: 275,
                  //width: 300,
                  child: Scrollbar(
                    child: ListView.separated(
                      scrollDirection: Axis.horizontal,
                      itemCount: snapshot.data["cnt"],
                        separatorBuilder: (context, index){
                          return VerticalDivider(
                            width: 5,
                            color: Colors.transparent,
                          );
                        },
                        itemBuilder: (BuildContext context, int index){
                        return Container(
                          decoration: utilities.containerBoxDecoration(),
                          width: 300,
                          child: Column(
                            children: <Widget>[
                              Text("${utilities.unixToNormal(snapshot.data["list"][index]["dt"])} (${DateTime.fromMillisecondsSinceEpoch(snapshot.data["list"][index]["dt"]).toUtc().timeZoneName})",style: utilities.moreInfoTextStyle().merge(TextStyle(fontSize: 15)),),
                              Padding(padding: EdgeInsets.all(10),),
                              Image.network('http://openweathermap.org/img/w/${snapshot.data["list"][index]["weather"][0]["icon"].toString()}.png',
                                height: 35,
                              ),
                              Text(snapshot.data["list"][index]["weather"][0]["main"].toString(),style: utilities.moreInfoTextStyle(),),
                              Text("(${snapshot.data["list"][index]["weather"][0]["description"].toString()})",style: utilities.moreInfoTextStyle().merge(TextStyle(fontStyle: FontStyle.italic)),),
                              Padding(
                                padding: const EdgeInsets.fromLTRB(0,10,0,20),
                                child: Text("${snapshot.data["list"][index]["main"]["temp"].toString()}\u2103",style: utilities.moreInfoTextStyle(),),
                              ),
                              Container(
                                width: 275,
                                child: Column(
                                  children: <Widget>[
                                    Row(
                                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                      children: <Widget>[
                                        Text("Min : ${snapshot.data["list"][index]["main"]["temp_min"].toString()}\u2103",style: utilities.moreInfoTextStyle(),),
                                        Text("Max : ${snapshot.data["list"][index]["main"]["temp_max"].toString()}\u2103",style: utilities.moreInfoTextStyle(),),
                                      ],
                                    ),
                                    Padding(padding: EdgeInsets.all(1.5),),
                                    Row(
                                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                      children: <Widget>[
                                        Text("Humidity : ${snapshot.data["list"][index]["main"]["humidity"].toString()}%",style: utilities.moreInfoTextStyle(),),
                                        Text("Clouds : ${snapshot.data["list"][index]["clouds"]["all"].toString()}%",style: utilities.moreInfoTextStyle(),),
                                      ],
                                    ),
                                    Padding(padding: EdgeInsets.all(1.5),),
                                    Row(
                                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                                      children: <Widget>[
                                        Text("Wind : ${snapshot.data["list"][index]["wind"]["speed"].toString()}m/s, ${snapshot.data["list"][index]["wind"]["deg"].toString()}\u00b0",style: utilities.moreInfoTextStyle(),),
                                        //Text("Wind dir : ",style: utilities.moreInfoTextStyle(),),
                                      ],
                                    ),
                                  ],
                                ),
                              )

                            ],
                          ),
                        );
                    }),
                  ),
                )
              ],
            ),
            floatingActionButton: FloatingActionButton(
                backgroundColor: Colors.black26,
                onPressed: () {
                  Navigator.pop(context);
                },
                child: Icon(Icons.chevron_left,color: utilities.textColor(),)
            ),
          ),
        );
      }else {
      return Container(
        child: Scaffold(
          backgroundColor: Colors.transparent,
          appBar: AppBar(
            backgroundColor: utilities.appBarColor(),
            title: Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: <Widget>[
                Text(
                  "Forecast",
                  style: utilities.containerTextStyle(),
                ),
                FlatButton(
                  child: Icon(Icons.refresh,color: utilities.textColor(),),
                  onPressed:(){
                    refresh(location);
                  },
                )
              ],
            ),
          ),
          body : Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              child: Container(
                  decoration: utilities.containerBoxDecoration(),
                  padding: EdgeInsets.all(10),
                  child: snapshot.data==null?Text("Loading...",style: utilities.moreInfoTextStyle().merge(TextStyle(letterSpacing: 10)),):Text(
                    snapshot.data.toString(),
                    style: utilities.moreInfoTextStyle(),
                  )),
            ),
          ),
          floatingActionButton: FloatingActionButton(
              backgroundColor: Colors.white12,
              onPressed: () {
                Navigator.pop(context);
              },
              child: Icon(Icons.chevron_left,color: utilities.textColor(),)
          ),
        ),
      );
    }
      },
    );

  }
}


lib/ui/more_info.dart
// MORE INFO PAGE

import 'package:flutter/material.dart';
import '../utilities.dart' as utilities;
import 'home.dart';

class MoreInfo extends StatefulWidget {
  final String location;
  MoreInfo({Key key, @required this.location}) : super(key: key);
  @override
  _MoreInfoState createState() => _MoreInfoState(location: location);
}

class _MoreInfoState extends State<MoreInfo> {
  final String location;
  var weatherInfo;
  _MoreInfoState({Key key, @required this.location});
  @override
  void initState(){
    super.initState();
    weatherInfo=getWeatherInfo(utilities.appId, location);
  }
  refresh(String loco){
    print("From refresh fn-$loco");
    setState(() {
      weatherInfo=getWeatherInfo(utilities.appId, loco);
    });
  }
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: weatherInfo,
      builder: (BuildContext context, AsyncSnapshot<Map> snapshot) {
        if (snapshot.hasData && snapshot.data.containsKey("main")) {
          return Container(
            decoration: BoxDecoration(
              image: DecorationImage(
                image: NetworkImage(
                    "https://images.pexels.com/photos/2516653/pexels-photo-2516653.jpeg?h=1366&w=768"),
                fit: BoxFit.cover,
              ),
            ),
            child: Scaffold(
              backgroundColor: Colors.black26,
              appBar: AppBar(
                backgroundColor: utilities.appBarColor(),
                title: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: <Widget>[
                    Text(
                      location == ""
                          ? "Please enter location"
                          : location == null ? utilities.defaultLocation : location,
                      style: utilities.containerTextStyle(),
                    ),
                    FlatButton(
                      child: Icon(Icons.refresh,color: utilities.textColor(),),
                      onPressed:(){
                        refresh(location);
                      },
                    )
                  ],
                ),
              ),
              body :ListView(
                children: <Widget>[
                  Padding(
                    padding: const EdgeInsets.all(25.0),
                    child: Container(
                      alignment: Alignment.topCenter,
                      child: Text(
                        "${snapshot.data["main"]["temp"].toString()}\u2103",
                        style: TextStyle(fontSize: 45)
                            .merge(utilities.containerTextStyle()),
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.fromLTRB(15,0,15,0),
                    child: Container(
                      decoration: utilities.containerBoxDecoration(),
                      padding: EdgeInsets.all(8),
                      child: Column(
                        children: <Widget>[
                          Padding(
                            padding: const EdgeInsets.fromLTRB(0,0,0,10),
                            child: Center(
                              child: Column(
                                mainAxisAlignment: MainAxisAlignment.center,
                                children: <Widget>[
                                  Image.network('http://openweathermap.org/img/w/${snapshot.data["weather"][0]["icon"]}.png',
                                    //height: 35,)
                                  ),
                                  Text("${snapshot.data["weather"][0]["main"].toString()}",style: utilities.moreInfoTextStyle().merge(TextStyle(fontSize: 22)),),

                                ],
                              ),
                            ),
                          ),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: <Widget>[
                              Text("Min : ${snapshot.data["main"]["temp_min"].toString()}\u2103",style: utilities.moreInfoTextStyle(),),
                              Text("Max : ${snapshot.data["main"]["temp_max"].toString()}\u2103",style: utilities.moreInfoTextStyle(),),
                            ],
                          ),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: <Widget>[
                              Text("Wind : ${snapshot.data["wind"]["speed"].toString()}m/s, ${snapshot.data["wind"]["deg"].toString()}\u00b0",style: utilities.moreInfoTextStyle(),),
                              Text("Clouds : ${snapshot.data["clouds"]["all"].toString()}%",style: utilities.moreInfoTextStyle(),),
                            ],
                          ),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: <Widget>[
                              Text(snapshot.data["visibility"]==null?"Visibility : N.A.":"Visibility : ${snapshot.data["visibility"].toString()}m",style: utilities.moreInfoTextStyle(),),
                              Text("Pressure : ${snapshot.data["main"]["pressure"].toString()}hPa",style: utilities.moreInfoTextStyle(),),
                            ],
                          ),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: <Widget>[
                              Text("Sunrise : ${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunrise"]*1000).toUtc().hour}:${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunrise"]*1000).toUtc().minute} ${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunrise"]*1000).toUtc().timeZoneName}",style: utilities.moreInfoTextStyle(),),
                              Text("Sunset : ${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunset"]*1000).toUtc().hour}:${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunrise"]*1000).toUtc().minute} ${DateTime.fromMillisecondsSinceEpoch(snapshot.data["sys"]["sunrise"]*1000).toUtc().timeZoneName}",style: utilities.moreInfoTextStyle(),),
                            ],
                          ),
                          Padding(
                            padding: EdgeInsets.all(10),
                          )
                        ],
                      ),
                    ),
                  ),
                ],
              ),
              floatingActionButton: FloatingActionButton(
                  backgroundColor: Colors.black26,
                  onPressed: () {
                    Navigator.pop(context);
                  },
                  child: Icon(Icons.chevron_left,color: utilities.textColor(),)
              ),
            ),
          );
        } else {
          return Container(
            child: Scaffold(
              backgroundColor: Colors.transparent,
              appBar:AppBar(
                backgroundColor: utilities.appBarColor(),
                title: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: <Widget>[
                    Text(
                      location == ""
                          ? "Please enter location"
                          : location == null ? utilities.defaultLocation : location,
                      style: utilities.containerTextStyle(),
                    ),
                    FlatButton(
                      child: Icon(Icons.refresh,color: utilities.textColor(),),
                      onPressed:(){
                        refresh(location);
                      },
                    )
                  ],
                ),
              ),
              body : Padding(
                padding: const EdgeInsets.all(8.0),
                child: Center(
                  child: Container(
                      decoration: utilities.containerBoxDecoration(),
                      padding: EdgeInsets.all(10),
                      child: snapshot.data==null?Text("Loading...",style: utilities.moreInfoTextStyle().merge(TextStyle(letterSpacing: 10)),):Text(
                        snapshot.data.toString(),
                        style: utilities.moreInfoTextStyle(),
                      )),
                ),
              ),
              floatingActionButton: FloatingActionButton(
                  backgroundColor: Colors.white12,
                  onPressed: () {
                    Navigator.pop(context);
                  },
                  child: Icon(Icons.chevron_left,color: utilities.textColor(),)
              ),
            ),
          );
        }
      },
    );
  }
}

lib/ui/home.dart
import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:weather_app/ui/forecast.dart';
import 'package:weather_app/utilities.dart' as utilities;
import 'more_info.dart';

class Home extends StatefulWidget {
  @override
  HomeState createState() => HomeState();
}

class HomeState extends State<Home> {
  String location = utilities.defaultLocation;
  var getLocation = TextEditingController();
  updateLocation() {
    setState(() {
      location = getLocation.text;
    });
  }
  var weatherInfo;
  @override

  void initState() {
    //We will get the default location (if set) here!
    super.initState();
    loadDefaultLocation();
  }

  loadDefaultLocation() async{
    SharedPreferences savedLocation=await SharedPreferences.getInstance();
    if(savedLocation.getString("location")!=null && savedLocation.getString("location")!="")/*Will be "" during first install*/{
      setState(() {
        location=savedLocation.getString("location");
      });
    }
    weatherInfo=getWeatherInfo(utilities.appId, location);
  }

  Widget build(BuildContext context) {
    return Container(
      decoration: BoxDecoration(
          image: DecorationImage(
        image: AssetImage(
            "images/blue-iphone-wallpaper-lock-screen-wallpaper-40465.jpg"),
        fit: BoxFit.cover,
      )),
      child: Scaffold(
        backgroundColor: Colors.transparent,
        appBar: AppBar(
          backgroundColor: utilities.appBarColor(),
          title: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: <Widget>[
              Text(
                "Weather",
                style: utilities.containerTextStyle(),
              ),
              Padding(
                padding: const EdgeInsets.fromLTRB(0, 5, 0, 5),
                child: Container(
                  decoration: utilities.containerBoxDecoration(),
                  width: 150,
                  child: TextField(
                    onEditingComplete: () {
                      //debugPrint(getLocation.text);
                      FocusScope.of(context).unfocus();
                      updateLocation();
                      refresh(location);
                    }, //To remove keyboard after typing...
                    textCapitalization: TextCapitalization.words,
                    //Styling
                    style: utilities.containerTextStyle(),
                    cursorColor: utilities.textColor(),
                    controller: getLocation,
                    decoration: InputDecoration(
                      hintText: "Location",
                      hintStyle: utilities.containerTextStyle(),
                      prefixIcon: Icon(
                        Icons.search,
                        color: utilities.textColor(),
                      ),
                      border: InputBorder.none,
                    ),
                  ),
                ),
              ),
              //Icon(Icons.menu)
            ],
          ),
        ),
        body: Builder(
          builder:(BuildContext context){
            return ListView(
            children: <Widget>[
              Container(
                  alignment: Alignment.topCenter,
                  padding: EdgeInsets.all(40),
                  child: Text(
                    location == ""
                        ? "Please enter location"
                        : location == null ? utilities.defaultLocation : location,
                    style: TextStyle(
                      //fontWeight: FontWeight.bold,
                      fontSize: 35,
                    ).merge(utilities.containerTextStyle()),
                    textAlign: TextAlign.center,
                  )),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container( //Weather Info set
                    child: FutureBuilder( //change_log 20-Oct-2019
                        future: weatherInfo,
                        builder: (BuildContext context, AsyncSnapshot<Map> snapshot) {
                          if (snapshot.hasData && snapshot.data.containsKey("main")) {
                            return Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              children: <Widget>[
                                Image.network(
                                  'http://openweathermap.org/img/w/${snapshot.data["weather"][0]["icon"]}.png',
                                  height: 35,
                                ),
                                Text(
                                  "It's Currently ${snapshot.data["main"]["temp"].toString()}\u{2103} at ${snapshot.data["name"].toString()}",
                                  style: TextStyle(
                                      fontSize: 25,
                                      fontWeight: FontWeight.w300,
                                      color: utilities.textColor()),
                                  textAlign: TextAlign.center,
                                ),
                                Container(
                                  padding: EdgeInsets.all(8),
                                  alignment: Alignment.center,
                                  child: Text(
                                    "Min : ${snapshot.data["main"]["temp_min"].toString()}\u{2103}\n"
                                        "Max : ${snapshot.data["main"]["temp_max"].toString()}\u{2103}\n"
                                        "Humidity : ${snapshot.data["main"]["humidity"].toString()}%",
                                    style: utilities.containerTextStyle(),
                                    textAlign: TextAlign.center,
                                  ),
                                ),
                                Row(
                                  children: <Widget>[
                                    FlatButton(
                                      child: Icon(Icons.refresh,color: utilities.textColor(),),
                                      color: Colors.transparent,
                                      onPressed: (){
                                        debugPrint("From FlatButton-$location");
                                        refresh(location);
                                      },
                                    ),
                                  ],
                                  mainAxisAlignment: MainAxisAlignment.end,
                                )
                              ],
                            );
                          } else {
                            return Center(
                              child: Container(
                                  padding: EdgeInsets.all(10),
                                  child: snapshot.data==null?Text("Loading...",style: utilities.moreInfoTextStyle().merge(TextStyle(letterSpacing: 10)),):Text(
                                    snapshot.data.toString(),
                                    style: utilities.containerTextStyle(),
                                  )),
                            );
                          }
                        }),

                    alignment: Alignment.topCenter,
                    padding: EdgeInsets.all(10),
                    decoration: utilities.containerBoxDecoration()),
              ),
              //BUTTONS//
              Padding(
                padding: EdgeInsets.all(50),
              ),
              Container(
                child: Column(
                  children: <Widget>[
                    ListTile(
                      title: FlatButton(
                        color: Colors.black26,
                        onPressed: () {
                          Navigator.push(context,
                              MaterialPageRoute(builder: (context) {
                            return MoreInfo(
                              location: location,
                            );
                          }));
                        },
                        child: Text(
                          "More info",
                          style: utilities.containerTextStyle(),
                        ),
                      ),
                    ),
                    ListTile(
                      title: FlatButton(
                        color: Colors.black26,
                        onPressed: () {
                          Navigator.push(context,
                              MaterialPageRoute(builder: (context) {
                                return Forecast(
                                  location: location,
                                );
                              }));
                        },
                        child: Text(
                          "Forecast",
                          style: utilities.containerTextStyle(),
                        ),
                      ),
                    ),
                    ListTile(
                      title: FlatButton(
                        color: Colors.black26,
                        onPressed: () {
                          utilities.setDefaultLocation(getLocation.text);
                          Scaffold.of(context).showSnackBar(SnackBar(
                            content: Text("Set!",style: utilities.containerTextStyle(),textAlign: TextAlign.center,),
                            backgroundColor: Colors.transparent,
                          ));
                        },
                        child: Text(
                          "Set location as default",
                          style: utilities.containerTextStyle(),
                        ),
                      ),
                    ),
                  ],
                ),
              )
            ],
          );},
        ),
      ),
    );
  }

  refresh(String loco){
    print("From refresh fn-$loco");
    setState(() {
      weatherInfo=getWeatherInfo(utilities.appId, loco);
    });
  }
}

/*Widget locationWeatherInfo(String appId, String location) {
  return
}*/

Future<Map> getWeatherInfo(String appId, String location) async {
  String apiURL =
      "http://api.openweathermap.org/data/2.5/weather?q=$location&APPID=$appId&units=metric";
  http.Response response = await http.get(apiURL);
  return jsonDecode(response.body);
}

Future<Map> getForecastInfo(String appId, String location) async {
  String apiURL =
      "https://api.openweathermap.org/data/2.5/forecast?q=$location&appid=$appId&units=metric";
  http.Response response = await http.get(apiURL);
  return jsonDecode(response.body);
}


